libvast/builtins/operators/from_read.cpp

```diff
@@ -0,0 +1,257 @@
+            co_yield x;
+          }
+        },
+        std::move(*loader));
+    } else {
+      VAST_ASSERT(parser_plugin_);
+      // TODO: Implement.
+      return caf::make_error(ec::unimplemented,
+                             "parser_plugin does not exist yet");
+    }
+  }
+
+  auto to_string() const -> std::string override {
+    if (loader_plugin_) {
+      return fmt::format("from {}", loader_plugin_->name());
+    } else {
+      VAST_ASSERT(parser_plugin_);
+      // This can not be round-tripped.
+      return fmt::format("<implicit from for {}>", parser_plugin_->name());
+    }
+  }
+
+private:
+  const loader_plugin* loader_plugin_{nullptr};
+  const plugin* parser_plugin_{nullptr};
+};
+
+class read_operator final : public crtp_operator<read_operator> {
+public:
+  explicit read_operator(const loader_plugin& loader)
+    : loader_plugin_{&loader} {
+  }
+
+  explicit read_operator(const plugin& parser) : parser_plugin_{&parser} {
+  }
+
+  auto
+  operator()(generator<chunk_ptr> input, operator_control_plane& ctrl) const
+    -> caf::expected<generator<table_slice>> {
+    if (parser_plugin_) {
+      VAST_ASSERT(!loader_plugin_);
+      // TODO: Implement.
+      return caf::make_error(ec::unimplemented,
+                             "parser_plugin does not exist yet");
+    } else {
+      VAST_ASSERT(loader_plugin_);
+      auto parser = loader_plugin_->make_default_parser({}, ctrl);
+      if (!parser) {
+        return parser.error();
+      }
+      // TODO: Reconsider this.
+      return std::invoke(
+        [](loader_plugin::parser parser,
+           generator<chunk_ptr> input) -> generator<table_slice> {
+          for (auto&& x : parser(std::move(input))) {
+            co_yield x;
+          }
+        },
+        std::move(*parser), std::move(input));
+    }
+  }
+
+  auto to_string() const -> std::string override {
+    if (parser_plugin_) {
+      return fmt::format("read {}", parser_plugin_->name());
+    } else {
+      VAST_ASSERT(loader_plugin_);
+      // This can not be round-tripped.
+      return fmt::format("<implicit read for {}>", loader_plugin_->name());
+    }
+  }
+
+private:
+  const plugin* parser_plugin_{nullptr};
+  const loader_plugin* loader_plugin_{nullptr};
+};
+
+class from_plugin final : public virtual operator_plugin {
+public:
+  caf::error initialize(const record&, const record&) override {
+    return {};
+  }
+
+  [[nodiscard]] std::string name() const override {
+    return "from";
+  };
+
+  auto make_operator(std::string_view pipeline) const
+    -> std::pair<std::string_view, caf::expected<operator_ptr>> override {
+    using parsers::optional_ws_or_comment, parsers::end_of_pipeline_operator,
+      parsers::identifier, parsers::required_ws_or_comment;
+    const auto* f = pipeline.begin();
+    const auto* const l = pipeline.end();
+    const auto p = optional_ws_or_comment >> identifier
+                   >> -(required_ws_or_comment >> string_parser{"read"}
```

Comment:
The operator name is already parsed at this point (this is because we have to dispatch to the `operator_plugin` with the same name). Hence we check whether we have (`from`) `X | ...` or (`from`) `X read Y`.